Leapp - New role and permissions, API route refactoring

stejskalleos opened the pull request.

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,17 @@
 module Api
   module V2
     class PreupgradeReportsController < ::Api::V2::BaseController
-      before_action :find_resource, only: %i[show]
```
Why?

stejskalleos replied:
Because now `@preupgrade_report` is searched by `find_by(job_invocation_id: params[:id])`, not by `find(params[:id])`.

adamruzicka replied:
Right, thanks

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,17 @@
       param :id, :identifier, required: true
-      def show; end
+      def show
+        @preupgrade_report = PreupgradeReport.find_by(job_invocation_id: params[:id])
```
Shouldn't this check if the user has permissions to see the preupgrade report?

stejskalleos replied:
Permission is already checked on controller level.

adamruzicka replied:
In REX we still check if the user has permissions to view a specific host when looking at job details (see https://github.com/theforeman/foreman_remote_execution/blob/master/app/controllers/job_invocations_controller.rb#L56). Would it make sense to limit what preupgrade report entries the user can see based on `view_hosts` permission?

@ares opinions?

ares replied:
yeah that makes sense, this is the case where we need to check that user has permission for both view_job_invocation and view_hosts I think

ares commented:
I spoke with @stejskalleos briefly about this, I think we can reuse REX permissions, in UI you need view_job_invocation anyway. I assume we just redefine the required permission in the controller. Since we depend on REX plugin, it's safe to assume such permissions exist.

stejskalleos commented:
Rebased & Updated.
* Removed Leapp role
* now just requiring `:view_job_invocations` permission from REX role.

stejskalleos commented:
Moving back to WIP, I need to make some changes for the integration with Hammer Leapp plugin - the show action for reports have to be under job invocation controller.

stejskalleos commented:
No need to do any changes, fixed in https://github.com/stejskalleos/hammer-cli-foreman-leapp/pull/2

adamruzicka commented on lib/tasks/foreman_leapp_tasks.rake:
```diff
@@ -32,6 +32,7 @@ namespace :foreman_leapp do
         task.patterns = ["#{ForemanLeapp::Engine.root}/app/**/*.rb",
                          "#{ForemanLeapp::Engine.root}/lib/**/*.rb",
                          "#{ForemanLeapp::Engine.root}/test/**/*.rb"]
+        task.options = ["--safe-auto-correct"]
```
Was this intentional?

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,32 @@
       param :id, :identifier, required: true
-      def show; end
+      def show
+        @preupgrade_report = PreupgradeReport.find_by(job_invocation_id: params[:id])
```
A single job can have multiple preupgrade reports (N reports for N hosts in the job), right? In that case, this cannot use `find_by` which returns a single record.

We should probably have `/job_invocations/:job_id/preupgrade_reports` which shows all the preupgrade reports in the job and `/job_invocations/:job_id/preupgrade_reports/:host_id` which shows a single preupgrade report for a given host in the job

stejskalleos replied:
Yeah that's true, I completely forgot that. How about splitting it into two actions:
```
/job_invocations/:job_id/preupgrade_reports (get all reports for the job_invocation)
/preupgrade_reports/:id (show report)
```

adamruzicka replied:
Fine by me

adamruzicka commented on app/controllers/preupgrade_reports_controller.rb:
```diff
@@ -2,6 +2,14 @@
 class PreupgradeReportsController < ::Api::V2::BaseController
   def index
-    @preupgrade_reports = resource_scope.search_for(*search_options)
+    @preupgrade_reports = PreupgradeReport.search_for(*search_options)
```
This should check `view_host` permission as well

stejskalleos replied:
But why? We just list preupgrade records, no hosts. For `Api:V2::JobInvocationsController index` permission `:view_host` is also not checked.

adamruzicka replied:
> We just list preupgrade records, no hosts.

Each preupgrade report is directly tied to a single host. If the users can't see the host, they shouldn't be able to see its preupgrade report. So if someone calls the index, they should get reports only for the hosts they can see.

> For Api:V2::JobInvocationsController index permission :view_host is also not checked

Because a job has many hosts and the index endpoint doesn't give any host-specific information. 

>  Same for the index action in Api::V2::PreupgradeReportsController

I'd say it should check it as well

stejskalleos commented:
Updated and rebased, with more changes than I was expecting:
* Now checking `view_hosts` permission for all controllers
* Actions in `api/v2/preupgrade_reports` splitted into three actions:
```
api_preupgrade_reports GET  /api(/:apiv)/preupgrade_reports(.:format)                     api/v2/preupgrade_reports#index {:format=>"json", :apiv=>/v2/}
 api_preupgrade_report GET  /api(/:apiv)/preupgrade_reports/:id(.:format)                 api/v2/preupgrade_reports#show {:format=>"json", :apiv=>/v2/}
                       GET  /api(/:apiv)/job_invocations/:id/preupgrade_reports(.:format) api/v2/preupgrade_reports#job_invocation {:format=>"json", :apiv=>/v2/}
```
* `resource_scope_for_index` and `resource_scope` was returning empty results for non-admin users (due to overriding `path_to_authenticate`)
* Tests checking response status AND response content - To be sure that user won't get empty results
* ... Plus using non-admin user in tests.

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,34 @@
+      api :GET, '/job_invocations/:id/preupgrade_reports', N_('List Preupgrade reports for Job invocation')
+      param :id, :identifier, required: true
+      def job_invocation
+        @preupgrade_reports = PreupgradeReport.where(job_invocation_id: params[:id])
```
A job is run against hosts `A` and `B`. As a user with `view_job_invocations` and `view_hosts` scoped to host `A` I make a request to this endpoint? Will I be able to see preupgrade reports for both hosts?

stejskalleos replied:
This should solve the issue:
```
hosts = JobInvocation.find(params[:id]).targeting.hosts.authorized(:view_hosts, Host)
@preupgrade_reports = PreupgradeReport.where(job_invocation_id: params[:id], host_id: hosts.ids)
```

Question is do we want to do this check also for `:show` and `:index` action?

adamruzicka replied:
I would say yes. The users should be able to see only what they're allowed to see

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,39 @@
       param :id, :identifier, required: true
-      def show; end
+      def show
+        @preupgrade_report = PreupgradeReport.find(params[:id])
+        not_found unless @preupgrade_report.host.authorized?(:view_hosts)
```
If you use the `joins().merge()` method in here, `find` should raise a `RecordNotFound` exception, which would render a 404

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,39 @@
+      api :GET, '/job_invocations/:id/preupgrade_reports', N_('List Preupgrade reports for Job invocation')
+      param :id, :identifier, required: true
+      def job_invocation
+        hosts = JobInvocation.find(params[:id]).targeting.hosts.authorized(:view_hosts, Host)
+        @preupgrade_reports = PreupgradeReport.where(job_invocation_id: params[:id], host_id: hosts.ids)
```
This probably works, but it could be simplified by using `.joins().merge()` without scoping the hosts to a job invocation.

adamruzicka commented on app/controllers/preupgrade_reports_controller.rb:
```diff
@@ -1,7 +1,21 @@
+    @preupgrade_reports = PreupgradeReport.search_for(*search_options)
+                                          .includes(:preupgrade_report_entries)
+                                          .joins(:host)
+                                          .merge(Host.all.authorized(:view_hosts, Host))
```
`.all` shouldn't be needed

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,41 @@
       param_group :search_and_pagination, ::Api::V2::BaseController
       def index
-        @preupgrade_reports = resource_scope_for_index
+        @preupgrade_reports = PreupgradeReport.search_for(*search_options)
```
`resource_scope_for_index` sets a couple of instance variables which are then used when rendering the response

```
$ curl -u Foo:changeme http://localhost:3000/api/v2/preupgrade_reports/ | jq                                                
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current                                                                                                                                                                                                                                             
                                 Dload  Upload   Total   Spent    Left  Speed
100  3146    0  3146    0     0   9987      0 --:--:-- --:--:-- --:--:--  9987
{                    
  "total": 0,                    
  "subtotal": 0,     
  "page": 1,                  
  "per_page": 20,                             
  "search": null,                            
  "sort": {                                    
    "by": null,    
    "order": null                              
  },                                                                                                                                                          
  "results": [                                                                 
    {                                                                        
      "id": 32,                                                               
      "host_id": 8,           
      "reported_at": "2020-05-04 14:28:07 UTC",
      "status": 0,           
      "metrics": {},                           
      "type": "PreupgradeReport",
      "origin": null,                          
      "job_invocation_id": 158   
    },
    ...
  }
]

I believe that not using this (and therefore not setting the variables) is the reason why `total` shows up as 0, even though there are some results
```

adamruzicka commented on app/controllers/api/v2/preupgrade_reports_controller.rb:
```diff
@@ -3,17 +3,41 @@
+      def job_invocation
+        @preupgrade_reports = PreupgradeReport.where(job_invocation_id: params[:id])
+                                              .joins(:host)
+                                              .merge(Host.authorized(:view_hosts, Host))
```
I'd say this should use the `index` layout.

adamruzicka commented:
Thank you @stejskalleos !
